ARCHITECTURE.md

```diff
@@ -0,0 +1,134 @@
+
+### The Client Object
+The main Bugsnag object that will be used in applications for the purpose of catching and notify of errors is the Client object.  This provides an API for the user to call on for the majority of the functions they will seek to utilise in the library, including: `notify`, `notify-exception`, `notify-error`, `leave-breadcrum`, and `deploy`.
+
+The client object has three constructor arguments that modify the way it will operate:
+- `configuration` Accepts a `Configuration` object for the user to customize behaviour of the Bugsnag notifier. The configuration options will be set in the most appropriate method for the framework being used, and parsed into this format.
+- `resolver` Accepts an object that implements the `ResolverInterface`.  This object will be responsible for returning an object that implements the `RequestInterface`, which will need to populate a `Report` upon request.  By default if not given a resolver the client will create a `BasicResolver`.
+- `guzzle` Accepts a `Guzzle` object provided by the guzzler library to use as an HTTP request object in the `HTTPClient` object.  By default the client will create one using the static `ENDPOINT` variable.  If it's necessary to change these defaults the static `make_guzzle` may be used with a defined base url and options array, with the result being passed into this argument.
+
+When utilising this library in a framework specific way it is advised that you wrap the client object creation and access in a class that correctly ties into the specific framework.  For example, in the [Bugsnag-laravel](https://github.com/bugsnag/bugsnag-laravel) library the client is wrapped in a `ServiceProvider` class to allow the Laravel framework to integrate it correctly.
+
+There is a provided static `make` method that creates the client object with suitable default `Configuration` and `Guzzle` arguments with the `api_key` and `endpoint` allowing the user to set their Bugsnag settings.
+
+Another important aspect of the client objects construction is the registering of callbacks with a `pipeline` object.  If created with the `make` method the default callbacks will be added to the `pipeline` unless
+the `defaults` argument is set to false.  These default callbacks will be responsible for populating the `report` object with information when necessary, and will should be customised to extract the most relevant information from the error and `resolver` objects used in the framework.
+
+### The Handler Class
+The handler class provides static functions for registering the Bugsnag unhandled error and exception handlers, as well as a shutdown handler.  This is separate to the client object as these handlers and methods of registering them will change across frameworks. 
+
+The static registration functions take an instance of the client object to reference in the event of a an exception by creating a `Report` object with it.
```

Comment:
What does "to reference" mean in this context? This is probably down to the fact that I know very little PHP :)